File: Networking/Requests/ResponseInfo.vb
Imports Networking.Data

Namespace Networking.Requests
    Public Class ResponseInfo
        Implements IMessage
        Public Sub New()
        End Sub

        Public request As RequestInfo
        Public manager As RequestManager

        Public sentAt As Date
        Public receivedAt As Date

        Public response As Object
        Public isSuccess As Boolean
        Public id As String

        Public Sub Deserialize(reader As Reader) Implements IDeserialize.Deserialize
            sentAt = reader.ReadDate()
            receivedAt = Date.Now

            id = reader.ReadCleanString()

            isSuccess = reader.ReadBool()

            response = reader.ReadAnonymous()
        End Sub

        Public Sub Serialize(writer As Writer) Implements ISerialize.Serialize
            writer.WriteDate(Date.Now)
            writer.WriteString(id)
            writer.WriteBool(isSuccess)
            writer.WriteAnonymous(response)
        End Sub
    End Class
End Namespace

File: Networking/Requests/ResponseInfo.cs
using Common.IO.Data;

using System;

namespace Networking.Requests
{
    public struct ResponseInfo : IData
    {
        public string Id { get; private set; }

        public object Response { get; private set; }

        public bool IsSuccess { get; private set; }

        public DateTime SentAt { get; private set; }
        public DateTime ReceivedAt { get; private set; }

        public RequestManager Manager { get; internal set; }

        public ResponseInfo(string id, object response, bool isSuccess)
        {
            Id = id;
            Response = response;
            IsSuccess = isSuccess;
            SentAt = DateTime.Now;
        }

        public void Deserialize(DataReader reader)
        {
            Id = reader.ReadString();
            Response = reader.ReadObject();
            SentAt = reader.ReadDate();
            IsSuccess = reader.ReadBool();
            ReceivedAt = DateTime.Now;
        }

        public void Serialize(DataWriter writer)
        {
            writer.WriteString(Id);
            writer.WriteObject(Response);
            writer.WriteDate(SentAt);
            writer.WriteBool(IsSuccess);
        }
    }
}